src/cunumeric/fft/fft.cu

```diff
@@ -0,0 +1,355 @@
+      CHECK_CUFFT(cufftXtExec(plan, (void*)in.ptr(zero), (void*)out.ptr(zero), (int)direction));
+  }
+};
+
+// For dimensions higher than 2D, we need to iterate through the input volume as 2D slices due to
+// limitations of cuFFT indexing in 1D
+template<typename OUTPUT_TYPE, typename INPUT_TYPE>
+struct cufft_axes_plan<3, OUTPUT_TYPE, INPUT_TYPE>{
+  __host__ static inline void execute(cufftHandle plan,
+                                      // AccessorWO<OUTPUT_TYPE, 3> out,
+                                      // AccessorRO<INPUT_TYPE,  3> in,
+                                      DeferredBuffer<OUTPUT_TYPE, 3>& out,
+                                      DeferredBuffer<INPUT_TYPE, 3>& in,
+                                      const Rect<3>& out_rect,
+                                      const Rect<3>& in_rect,
+                                      int axis,
+                                      fftDirection direction) {
+    bool is_inner_axis = (axis == 1);
+    if(is_inner_axis) {
+      // TODO: use PointInRectIterator<DIM>
+      auto num_slices = in_rect.hi[0] - in_rect.lo[0] + 1;
+      for(unsigned n = 0; n < num_slices; ++n){
+        const Point<3> offset = Point<3>(n, 0, 0);
+        CHECK_CUFFT(cufftXtExec(plan, (void*)in.ptr(offset), (void*)out.ptr(offset), (int)direction));
+      }
+    }
+    else {
+      const Point<3> zero   = Point<3>::ZEROES();
+      CHECK_CUFFT(cufftXtExec(plan, (void*)in.ptr(zero), (void*)out.ptr(zero), (int)direction));
+    }
+  }
+};
+
+// Perform the FFT operation as multiple 1D FFTs along the specified axes.
+// For now, it only supports up to 3D FFTs, but final plan is having support for
+// N-dimensional FFTs using this approach.
+template <int DIM, typename OUTPUT_TYPE, typename INPUT_TYPE>
+__host__ static inline void cufft_over_axes(AccessorWO<OUTPUT_TYPE, DIM> out,
+                                            AccessorRO<INPUT_TYPE, DIM> in,
+                                            const Rect<DIM>& out_rect,
+                                            const Rect<DIM>& in_rect,
+                                            std::vector<int64_t>& axes,
+                                            fftType type,
+                                            fftDirection direction)
+{
+    cudaStream_t stream;
+    CHECK_CUDA(cudaStreamCreateWithFlags(&stream, cudaStreamNonBlocking));
+
+    size_t workarea_size = 0;
+    dim_t n[DIM];
+    dim_t inembed[DIM];
+    dim_t onembed[DIM];
+
+    // Full volume dimensions / strides
+    const Point<DIM> zero   = Point<DIM>::ZEROES();
+    const Point<DIM> one    = Point<DIM>::ONES();
+    Point<DIM> fft_size_in  =  in_rect.hi -  in_rect.lo + one;
+    Point<DIM> fft_size_out = out_rect.hi - out_rect.lo + one;
+    size_t num_elements_in  = 1;
+    size_t num_elements_out = 1;
+    for(int i = 0; i < DIM; ++i) {
+      n[i]          = /*(type == fftType::FFT_R2C || type == fftType::FFT_D2Z) ? fft_size_in[i] :*/ fft_size_out[i];
```

Comment:
Why is this? Can you explain with a comment? I would have though we needed to keep `(type == fftType::FFT_R2C || type == fftType::FFT_D2Z) ? fft_size_in[i] : fft_size_out[i])`

Reply:
I forgot to remove that old check. The reason why it's not necessary is because R2C and C2R are done by a different function, `cufft_over_axis_r2c_c2r`. This one is only for C2C.

But I can see that it's confusing and this is not stated anywhere in this cufft_over_axes, so I'll update it to make it clearer 😄